### ISSUE TITLE
Changes to drop volatile lead to kokkos-kernels build errors with -Werror, gcc/8.3.0, gcc/7.2.0

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#4959]
@PhilMiller @dalg24 with #4951 merged I'm seeing compilation errors (-Werror=maybe-uninitialized) in kokkos-kernels with gcc/7.2.0,

```
In file included from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Internal.hpp:8:0,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_QR_WithColumnPivoting_TeamVector_Impl.hpp:7,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/KokkosBatched_QR_WithColumnPivoting_Decl.hpp:28,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/batched/dense/Test_Batched_TeamVectorQR_WithColumnPivoting.hpp:12,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/batched/dense/Test_Batched_Dense.hpp:85,
                 from /home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/unit_test/threads/Test_Threads_Batched_Dense.cpp:5:
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp: In static member function static int KokkosBatched::TeamVectorQR_WithColumnPivotingInternal::invoke(const MemberType&, int, int, ValueType*, int, int, ValueType*, int, IntType*, int, ValueType*, int&) [with MemberType = Kokkos::Impl::ThreadsExecTeamMember; ValueType = double; IntType = int]:
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:58:55: error: value.Kokkos::ValLocScalar::loc may be used uninitialized in this function [-Werror=maybe-uninitialized]
       Kokkos::single(Kokkos::PerTeam(member), [&]() { *idx = value.loc; });
                                                       ^
/home/jenkins/blake-new/workspace/KokkosKernels_PullRequest_GCC720/kokkos-kernels/src/batched/dense/impl/KokkosBatched_FindAmax_Internal.hpp:46:26: note: value.Kokkos::ValLocScalar::loc was declared here
       reducer_value_type value;
                          ^~~~~
```

Should I open a new issue or reopen this issue?

_Originally posted by @ndellingwood in https://github.com/kokkos/kokkos/issues/4941#issuecomment-1097053909_

**Comment by ndellingwood** [comment]
@PhilMiller - here is a reproducer with gcc/7.2.0, Threads backend, on Blake testbed

**Reproducer** - blake testbed, gcc/7.2.0

```
# Load modules
module load cmake/3.19.3 gcc/7.2.0

# Prepare Serial build
$KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Threads,Serial --arch=SKX --compiler=g++ --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wignored-qualifiers -Wempty-body -Wclobbered -Wuninitialized " --cxxstandard="14" --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --no-examples

# Compile kokkos-kernels
make -j8
```

**Comment by ndellingwood** [comment]
Addressed by kokkos/kokkos-kernels/#1383, passed PR testing and merged, thanks @dalg24 ! Closing issue
